### ISSUE TITLE
Add Holiday graphics

### ISSUE THREAD

**Issue by lorezyra** [lorezyra/Richie2.com#62]
Would like to add some graphics for specific holidays throughout the year.

The current hero graphic is perfect for the summer. Let's add some for the following:

- [x] My birthday (November 10)
- [x] Winter: Skiing
- [x] New Year's EVE
- [x] New Year's Day
- [x] National Foundation
- [x] Chinese Lunar New Year
- [x] Spring
- [x] Marine Day 
- [x] Mountain Day
- [x] Fall time
- [x] Birthday of Japan's Emperor
- [x] Golden Week

Basically, every holiday listed here:
* https://holidayapi.com/countries/jp/2021

- [x] This can be used for New Year's Day: https://stackoverflow.blog/2021/05/31/shipping-confetti-to-stack-overflows-design-system/?utm_source=Iterable&utm_medium=email&utm_campaign=the_overflow_newsletter

**Comment by Asem-prg** [comment]
Do you mean that you want to change the character in hero section according to specific holidays throughout the year?

**Comment by lorezyra** [comment]
Yeah, I do. However, let's not start on this `enhancement` right now. 
Let's comeback to this later. After we have the rest of the site done, we can look at this idea.

**Comment by lorezyra** [comment]
All "images" should be SVG embedded into the HTML.

**Comment by Asem-prg** [comment]
Would you still want to do this issue?

**Comment by lorezyra** [comment]
Yes, I'm interested. Could you create some Mock-ups of what each could look like in Adobe XD?

**Comment by Asem-prg** [comment]
Sure

**Comment by Asem-prg** [comment]
I have made an adobe Xd file to present how the character will look like, in addition of an illustrator file and SVG file for each character. Can you review it and give me your opinion, please?

**Comment by lorezyra** [comment]
**Love it!**

There are a few minor changes we should make:

- [ ] Winter Skiing:
    - Show my character with skis going down hill
- [ ] Chinese New year:
    - Add a Lantern with the kanji: 「新年快乐」("Happy New Year" in simplified Chinese )
    - ![新年](https://user-images.githubusercontent.com/1673222/145309464-b85bcd1c-3699-4d0c-87c4-3f01e2101418.png)
    - ![新年快乐](https://user-images.githubusercontent.com/1673222/145309541-2a746602-11af-4707-bb0b-c16727325aa7.png)
    - ![New Year lanterns](https://as2.ftcdn.net/v2/jpg/03/00/18/35/1000_F_300183520_QU4vFV8vC6uceSLyJb42EgnmxidCpr35.jpg)
- [ ] Mountain Day:
    - The micro-Mountain below my feet is odd. As we have a forest and mountain range in the background, we can remove the "mountain" below the character's feet.

If you want to use a stock image from Adobe Stock, let me know which one and I'll buy it for you.

**Comment by lorezyra** [comment]
Looks great!

**Comment by lorezyra** [comment]
Now, let's create an array of objects inside a `isHoliday()` function that swaps out the hero image with the holiday graphic.

Something like:
```JS
/**
 * @function isHoliday
 * @desc Determines if today is a known holiday for Japan (the US or China) and returns data if it is.
 * @param {}
 * @return {boolean} result - Returns `0` if not holiday; Otherwise, returns object 
 * @todo Check if `0` or `{object}`
 */
function isHoliday() {

    // holiday listed here: 
    //     https://holidayapi.com/countries/jp/2022

    let holiday = new Array();
    let today = new Date();
    let hIndex = -1;
    let result = false;
    const thisDay = (today.getMonth() + 1) + "/" + today.getDate(); // outputs string like "12/31"

    // let searchDay = function (myDay, hIndex, arr, tDay = thisDay) {
    //     // using custom function to better understand findIndex prototype
    //     // console.log(myDay);
    //     // console.log(hIndex);
    //     if (myDay[0] === tDay) {
    //         return (myDay[1]);
    //     } else {
    //         return false;
    //     }
    // };


    // [date string in format "M/D" (month / date), {object with image data)]
    holiday = [
        ["1/1", {
            'img': "~/path/to/img1.svg"
        }],
        ["2/1", {
            'img': "~/path/to/img2.svg"
        }],
        ["3/1", {
            'img': "~/path/to/img3.svg"
        }],
        ["4/1", {
            'img': "~/path/to/img4.svg"
        }],
        ["12/13", {
            'img': "~/path/to/img5.svg"
        }]
    ];

    // hIndex = holiday.findIndex(searchDay); 
    hIndex = holiday.findIndex( hDay => hDay[0] === thisDay );

    if (hIndex > -1) {
        result = holiday[hIndex][1];
    }

    return result;
    // holiday[4][1].img would return the path to graphic

}
```

In the `holiday` array, you can add any other data needed for that date to be included inside the object. This makes it function similar to how databases operate using an index function. I left the `searchDay()` function as it can help if we need to use more complex date strings. For example, the "3rd Thursday of November" could be used as the index and would require more advanced processing. But this could be simplified to a calculated date that the index initializes with...

## Usage:
```JS
isHoliday().img;
```

Feel free to modify this as you see fit.

**Comment by Asem-prg** [comment]
External SVG wouldn't work well on both [Day, Night] themes because it has a fixed color. My idea is to replace the current SVG in the `index.html` page with another SVG using JavaScript by calling a function in another `.js` file for each SVG, so we don't have to increase the size of one `.js` file.

How do you think?

**Comment by lorezyra** [comment]
Regarding the SVG colors with Day/Night themes, it's trivial to add CSS classes to the elements of the SVG and adjust the colors. I believe you did something similar on the error pages and on the hero's foot path (oval).

The default should always load the base (hero) SVG. After page loads, a script should check `isHoliday` and if not false, hide existing hero and load the holiday graphic (returned by `isHoliday`). You can easily add color profiles as data inside the object returned by the function. For the home page, please don't use the loading filter as you have with the other pages (like on the play page) as most users would close the page before it loaded. 

Workflow:

1. `index.html` loads
2. `isHoliday` runs
3. If holiday, load {img} from external path with {colorProfile} data
4. While img is loading, hide hero element
5. delete the hero element after {img} loads

You can add more data to the object within the array. For example:
```js
holiday = [
        ["1/1", {
            'img': "~/path/to/img0.svg",
           'colorProfile': ["CSS class data"],
           'cssDayClass': ["CSS classname"],
           'cssNightClass': ["CSS classname"]
        }],
        ["2/1", {
            'img': "~/path/to/img1.svg",
           'colorProfile': ["CSS class data"],
           'cssDayClass': ["CSS classname"],
           'cssNightClass': ["CSS classname"]
        }],
        ["3/1", {
            'img': "~/path/to/img2.svg",
           'colorProfile': ["CSS class data"],
           'cssDayClass': ["CSS classname"],
           'cssNightClass': ["CSS classname"]
        }],
        ["4/1", {
            'img': "~/path/to/img3.svg",
           'colorProfile': ["CSS class data"],
           'cssDayClass': ["CSS classname"],
           'cssNightClass': ["CSS classname"]
        }],
        ["12/13", {
            'img': "~/path/to/img4.svg",
           'colorProfile': ["CSS class data"],
           'cssDayClass': ["CSS classname"],
           'cssNightClass': ["CSS classname"]
        }]
    ];
```

**Comment by Asem-prg** [comment]
Hi Richie

I have done changing the characters based on Holidays. However, there are some holidays I couldn't find their date on "https://holidayapi.com/countries/jp/2021" in addition to your birthday.

[] ChineseLunarNewYear
[] Falltime
[] GoldenWeek
[] Spring
[] WinterSkiing

Could you tell me about them, please?

**Comment by Asem-prg** [comment]
Hi Richie

I have done changing the characters based on Holidays. However, there are some holidays I couldn't find their date on "https://holidayapi.com/countries/jp/2021" in addition to your birthday.

ChineseLunarNewYear 
Falltime 
GoldenWeek 
Spring 
WinterSkiing

Could you tell me about them, please?

**Comment by lorezyra** [comment]
- ChineseLunarNewYear
   - https://en.wikipedia.org/wiki/Chinese_New_Year
   - https://holidayapi.com/countries/cn/2022
   - For 2022: February 1st
   - Will need a method to calculate this day...
       - https://stackoverflow.com/questions/30719176/algorithm-to-find-the-gregorian-date-of-the-chinese-new-year-of-a-certain-gregor
       - [date-chinese](https://github.com/commenthol/date-chinese)
- Falltime
   - Reference to the Lunar equinox autumn event: about September 23
- GoldenWeek
   - From Showa day through Children's day (basically the 1st week of May)
- Spring
   - Reference to the Lunar equinox spring event: about March 21
- WinterSkiing
   - From Lunar solstice (December 21) ~ End of February


For Holidays that use Lunar events/calendars, we could adapt a JS Library from [SunCalc](https://github.com/mourner/suncalc).

**Comment by Asem-prg** [comment]
Could you review the latest updates, please?

**Comment by lorezyra** [comment]
Is it not faster to call a stored variable than to call the same function multiple times?
https://github.com/lorezyra/Richie2.com/blob/5f5667b1d18b4b40de16c22f3890c2a4af123602/assets/js/main.dev.js#L709

Perhaps define a new variable to store it?
```js 
let todayYear = today.getFullYear()
```

Furthermore, is it necessary to define the year? It's a safe assumption that the year will remain the same. For dates that cross a calendar year (like Winter Skiing), we could simply add another array element.
```js
            },
            {
                'startDate': 0102,
                'endDate': 0228,
                'img': "/assets/img/Holidays/WinterSkiing.svg",
            },
            {
                'startDate': 1221,
                'endDate':  1230,
                'img': "/assets/img/Holidays/WinterSkiing.svg",
            }];

```

Also, both the month & day must be a 2 digit number.  In fact, we could use a simple number in the format ``MMDD`` like "1231" to represent New Year's Eve. A number take far less memory than a string and is typically faster to access.

**Comment by Asem-prg** [comment]
The only reason I've used the year in the date is for `Winter Skiing` holiday, because it continued to February of the next year, If I didn't use the year for this holiday, all the days between "12-21" and "2-28" will be a holiday for `Winter Skiing` in the same year.

I liked your idea to use an integer number instead of string with this format `YYYYMMDD` I will try to use it.

I will do the rest of changes.

**Comment by lorezyra** [comment]
You still don't need the year. Simply use two entries as I suggested above.

**Comment by Asem-prg** [comment]
Okay.

**Comment by lorezyra** [comment]
Besides, the snow isn't all that great in the early season (December). And, we could use a separate image for the skiing after Jan01.

**Comment by Asem-prg** [comment]
Could you please review the latest updates?

**Comment by lorezyra** [comment]
Love the Winter Ski SVG updates!

**Comment by lorezyra** [comment]
Site looks awesome! Great work!!

**Comment by lorezyra** [comment]
Oh, Just noticed that we are missing a graphic for New Year's EVE. Can you add that?

**Comment by Asem-prg** [comment]
Sure, I will add it.

**Comment by lorezyra** [comment]
Sweetness! Looks awesome!
